sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/NonAsyncTaskShouldNotReturnNull.cs

```diff
@@ -0,0 +1,114 @@
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.VisualBasic;
+using Microsoft.CodeAnalysis.VisualBasic.Syntax;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using SonarAnalyzer.Helpers.VisualBasic;
+
+namespace SonarAnalyzer.Rules.VisualBasic
+{
+    [DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+    [Rule(DiagnosticId)]
+    public sealed class NonAsyncTaskShouldNotReturnNull : NonAsyncTaskShouldNotReturnNullBase
+    {
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
+
+        private static readonly ISet<SyntaxKind> PossibleEnclosingSyntax =
+            new HashSet<SyntaxKind>
+            {
+                SyntaxKind.VariableDeclarator,
+                SyntaxKind.PropertyBlock,
+                SyntaxKind.FunctionBlock
+            };
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var nullLiteral = (LiteralExpressionSyntax)c.Node;
+
+                    if (!IsParentReturnOrReturnTernary(nullLiteral))
+                    {
+                        return;
+                    }
+
+                    var enclosingMember = GetEnclosingMember(nullLiteral);
+                    if (enclosingMember != null &&
+                        !enclosingMember.IsKind(SyntaxKind.VariableDeclarator) &&
+                        IsInvalidEnclosingSymbolContext(enclosingMember, c.SemanticModel))
+                    {
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, nullLiteral.GetLocation()));
+                    }
+                },
+                SyntaxKind.NothingLiteralExpression);
+        }
+
+        private static bool IsParentReturnOrReturnTernary(SyntaxNode node)
+        {
+            var parent = node.GetFirstNonParenthesizedParent();
+
+            if (parent.IsKind(SyntaxKind.ReturnStatement))
+            {
+                return true;
+            }
+            else if (parent.IsKind(SyntaxKind.TernaryConditionalExpression))
+            {
+                var grandParent = parent.GetFirstNonParenthesizedParent();
+
+                return grandParent.IsKind(SyntaxKind.ReturnStatement);
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        private static SyntaxNode GetEnclosingMember(LiteralExpressionSyntax literal)
+        {
+            foreach (var ancestor in literal.Ancestors())
+            {
+                if (ancestor.IsKind(SyntaxKind.MultiLineFunctionLambdaExpression) ||
```

Comment:
Would it be clearer to have a  `switch (ancestor.Kind())` (and remove the hashset :) ) ?